### ISSUE TITLE
Queries are too complex

### ISSUE THREAD

**Issue by pdt256** [inklabs/kommerce-core#88]
**Do you have any ideas how to reduce the complexity?**

The current way to get a **ProductDTO**:

```php
try {
    $request = new GetProductRequest($productId);
    $response = new GetProductResponse($this->getPricing());
    $this->dispatchQuery(new GetProductQuery($request, $response));
} catch (EntityNotFoundException $e) {
    return abort(404);
}

$productDTO = $response->getProductDTO();
```

The classes required to satisfy the request:

[GetProductQuery](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/Action/Product/GetProductQuery.php)
[GetProductRequest](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/Action/Product/Query/GetProductRequest.php)
[GetProductResponseInterface](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/Action/Product/Query/GetProductResponseInterface.php)
[GetProductResponse](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/Action/Product/Query/GetProductResponse.php)
[GetProductHandler](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/ActionHandler/Product/GetProductHandler.php)

The **GetProductHandler** defers to the ProductService:

```php
class GetProductHandler {
    public function handle(GetProductQuery $query)
    {
        $product = $this->productService->findOneById(
            $query->getRequest()->getProductId()
        );
    }

```

The **ProductService** defers to the ProductRepository:

```php
class ProductService {
    public function findOneById(UuidInterface $id)
    {
         return $this->productRepository->findOneById($id);
    }
```

The ProductRepository defers to the Doctrine EntityRepository:

```php
class ProductRepository {
    public function findOneById(UuidInterface $id)
    {
        return $this->returnOrThrowNotFoundException(
            parent::findOneBy(['id' => $id])
        );
    }
```

**Comment by pdt256** [comment]
Commands are much less complex:

```php
$tagId = $request->input('tagId');
$productId = $request->input('productId');

try {
    $this->dispatch(new AddTagToProductCommand(
        $productId,
        $tagId
    ));

    $this->flashSuccess('Successfully added product');
} catch (KommerceException $e) {
    $this->flashError('Unable to add product');
}
```

Classes required:

[AddTagToProductCommand](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/Action/Product/AddTagToProductCommand.php)
[AddTagToProductHandler](https://github.com/inklabs/kommerce-core/blob/9a79db98d6ca76d2ab7f8e82b01cdfa02d2a724d/src/ActionHandler/Product/AddTagToProductHandler.php)

**Comment by pdt256** [comment]
**Before:**
```php
    public function index(Request $httpRequest)
    {
        $queryString = $httpRequest->query('q');

        $request = new ListWarehousesRequest(
            $queryString,
            $this->getPaginationDTO(20)
        );

        $response = new ListWarehousesResponse();
        $this->dispatchQuery(new ListWarehousesQuery($request, $response));

        $warehouses = $response->getWarehouseDTOs();
        $pagination = $response->getPaginationDTO();

        return $this->renderTemplate(
            '@admin/inventory/warehouse/index.twig',
            [
                'warehouses' => $warehouses,
                'pagination' => $pagination,
                'queryString' => $queryString,
            ]
        );
    }
```

**After:**
```php
    public function index(Request $httpRequest)
    {
        $queryString = $httpRequest->query('q');

        /** @var ListWarehousesResponse $response */
        $response = $this->dispatchQuery(new ListWarehousesQuery(
            $queryString,
            $this->getPaginationDTO(20)
        ));

        $warehouses = $response->getWarehouseDTOs();
        $pagination = $response->getPaginationDTO();

        return $this->renderTemplate(
            '@admin/inventory/warehouse/index.twig',
            [
                'warehouses' => $warehouses,
                'pagination' => $pagination,
                'queryString' => $queryString,
            ]
        );
    }
```

**Comment by pdt256** [comment]
37 to refactor:

$ find . -name *Query.php
- [x] Attribute/GetAttributeQuery.php
- [x] Attribute/GetAttributeValueQuery.php
- [x] Attribute/ListAttributesQuery.php
- [x] Cart/GetCartBySessionIdQuery.php
- [x] Cart/GetCartByUserIdQuery.php
- [x] Cart/GetCartQuery.php
- [x] CartPriceRule/GetCartPriceRuleQuery.php
- [x] CartPriceRule/ListCartPriceRulesQuery.php
- [x] CatalogPromotion/GetCatalogPromotionQuery.php
- [x] CatalogPromotion/ListCatalogPromotionsQuery.php
- [x] Configuration/GetConfigurationsByKeysQuery.php
- [x] Coupon/GetCouponQuery.php
- [x] Coupon/ListCouponsQuery.php
- [x] Option/GetOptionQuery.php
- [x] Option/ListOptionsQuery.php
- [x] Order/GetOrderItemQuery.php
- [x] Order/GetOrderQuery.php
- [x] Order/GetOrdersByUserQuery.php
- [x] Order/ListOrdersQuery.php
- [x] Product/GetProductQuery.php
- [x] Product/GetProductsByIdsQuery.php
- [x] Product/GetProductsByTagQuery.php
- [x] Product/GetRandomProductsQuery.php
- [x] Product/GetRelatedProductsQuery.php
- [x] Product/ListProductsQuery.php
- [x] Shipment/GetLowestShipmentRatesByDeliveryMethodQuery.php
- [x] Shipment/GetShipmentRatesQuery.php
- [x] Shipment/GetShipmentTrackerQuery.php
- [x] Shipment/ListAdHocShipmentsQuery.php
- [x] Tag/GetTagQuery.php
- [x] Tag/GetTagsByIdsQuery.php
- [x] Tag/ListTagsQuery.php
- [x] TaxRate/ListTaxRatesQuery.php
- [x] User/GetUserByEmailQuery.php
- [x] User/GetUserQuery.php
- [x] User/ListUsersQuery.php
- [x] Warehouse/ListWarehousesQuery.php

**Comment by pdt256** [comment]
Refactoring is really easy. Around 3 minutes per Query:

[![IMAGE ALT TEXT](http://img.youtube.com/vi/LbbR9MWs6cA/0.jpg)](http://www.youtube.com/watch?v=LbbR9MWs6cA "Video Title")

Code from the video: https://github.com/inklabs/kommerce-core/commit/66ddaf1f5876d713e21765f55e52e8ea4fa9ec7a

**Comment by pdt256** [comment]
Merged to master via #110
